### ISSUE TITLE
bctls-fips 1.0.16 -> 1.0.17 breaks TLS handshake

### ISSUE THREAD

**Issue by MatteoVH** [bcgit/bc-java#1526]
When attempting to upgrade bctls-fips to the latest version (1.0.17 from 1.0.16), I started encountering TLS handshake failures with the following stack trace and the connection fails to establish:

```
[org.bouncycastle.jsse.provider.ProvTlsClient.WARN] [ProvTlsClient.java:notifyAlertRaised:451] - [client #9202 @17559fcb] raised fatal(2) internal_error(80) alert: Failed to read record
java.lang.UnsupportedOperationException: null
	at java.base/javax.net.ssl.ExtendedSSLSession.getStatusResponses(ExtendedSSLSession.java:157)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:220)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:129)
	at org.bouncycastle.jsse.provider.ImportX509TrustManager_7.checkServerTrusted(ImportX509TrustManager_7.java:56)
	at org.bouncycastle.jsse.provider.ProvSSLSocketDirect.checkServerTrusted(ProvSSLSocketDirect.java:131)
	at org.bouncycastle.jsse.provider.ProvTlsClient$1.notifyServerCertificate(ProvTlsClient.java:377)
	at org.bouncycastle.tls.TlsUtils.processServerCertificate(TlsUtils.java:4849)
	at org.bouncycastle.tls.TlsClientProtocol.handleServerCertificate(TlsClientProtocol.java:797)
	at org.bouncycastle.tls.TlsClientProtocol.handleHandshakeMessage(TlsClientProtocol.java:687)
	at org.bouncycastle.tls.TlsProtocol.processHandshakeQueue(TlsProtocol.java:715)
	at org.bouncycastle.tls.TlsProtocol.processRecord(TlsProtocol.java:591)
	at org.bouncycastle.tls.RecordStream.readRecord(RecordStream.java:247)
	at org.bouncycastle.tls.TlsProtocol.safeReadRecord(TlsProtocol.java:879)
	at org.bouncycastle.tls.TlsProtocol.blockForHandshake(TlsProtocol.java:427)
	at org.bouncycastle.tls.TlsClientProtocol.connect(TlsClientProtocol.java:88)
	at org.bouncycastle.jsse.provider.ProvSSLSocketDirect.startHandshake(ProvSSLSocketDirect.java:426)
	at org.bouncycastle.jsse.provider.ProvSSLSocketDirect.handshakeIfNecessary(ProvSSLSocketDirect.java:519)
	at org.bouncycastle.jsse.provider.ProvSSLSocketDirect$AppDataOutput.write(ProvSSLSocketDirect.java:625)
```

I'm installing the Bouncycastle providers as position 1 and 2 as follows, both before and after the upgrade:
`Security.insertProviderAt(new BouncyCastleFipsProvider(), 1);`
`Security.insertProviderAt(new BouncyCastleJsseProvider("fips:BCFIPS"), 2);`

Running on `openjdk_11.0.20.1_1`, with `-Dorg.bouncycastle.fips.approved_only=true`

With the following bc versions:

```        
"bc-fips", "1.0.2.3"
"bcpg-fips", "1.0.7.1"
"bcpkix-fips", "1.0.7"
"bctls-fips", "1.0.17"
```

**Comment by peterdettman** [comment]
The first obvious issue is that you are not picking up the BCJSSE trust manager. Check that you have the following settings in your java.security config:

ssl.KeyManagerFactory.algorithm=PKIX
ssl.TrustManagerFactory.algorithm=PKIX

**Comment by MatteoVH** [comment]
Thank you Peter. I'm applying those settings and I will get back to you.

**Comment by peterdettman** [comment]
@MatteoVH Any updates?

**Comment by ericks-apptio** [comment]
@peterdettman FWIW I was having this same issue and was able to resolve it by adding the java.security settings you mentioned above

**Comment by MatteoVH** [comment]
I tried to apply those settings dynamically since we don't have an established `java.security` file and ran into the same exception as before, meaning they were likely not applied correctly. I haven't had time to investigate further yet, but plan to return to it soon.

Applying dynamically at the initialization of my app:
```
Security.setProperty("ssl.KeyManagerFactory.algorithm", "PKIX");
Security.setProperty("ssl.TrustManagerFactory.algorithm", "PKIX");
```

Good to hear that this has solved the issue for Evan.

**Comment by MatteoVH** [comment]
Hi Peter,

I was able to fix the problem. There were some cases where we were pulling "SunX509" instances of the Key or TrustManager Factories. I corrected that and SSL now works with v1.0.17 of bctls-fips.

Thank you for the help!
